### ISSUE TITLE
NuGet packages need to be restored before building

### ISSUE THREAD

**Issue by badrianiulian** [filipnavara/mac-playground#33]
Hello!
So I've been  trying to compile this project but I've hit a brick wall.
I get errors when the project is loaded regarding some NuGet packages:

```
Getting restore information for project /Users/user/Compile/mac-playground/FormsTest/FormsTest.csproj
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(109,5) : error NETSDK1139: The target platform identifier macos was not recognized.
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(109,5) : error NETSDK1139: The target platform identifier macos was not recognized.
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(109,5) : error NETSDK1139: The target platform identifier macos was not recognized.
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(109,5) : error NETSDK1139: The target platform identifier macos was not recognized.
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(109,5) : error NETSDK1139: The target platform identifier macos was not recognized.
Unable to create package spec for project. '/Users/user/Compile/mac-playground/FormsTest/FormsTest.csproj'
```

While browsing through Dependencies and manually updating NuGet packages, these are the errors I get:
System.Drawing.Common:
```
  GET https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/7d9f5c21-0d79-403f-bfe3-9a4506529760/nuget/v3/registrations2-semver2/system.drawing.common/index.json
  GET https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/registrations2-semver2/system.drawing.common/index.json
  GET https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/registrations2-semver2/system.drawing.common/index.json
  GET https://pkgs.dev.azure.com/azure-public/3ccf6661-f8ce-4e8a-bb2e-eff943ddd3c7/_packaging/be90bfa8-5c00-452b-ba9f-226ce4366992/nuget/v3/registrations2-semver2/system.drawing.common/index.json
  GET https://api.nuget.org/v3/registration5-gz-semver2/system.drawing.common/index.json
  OK https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/registrations2-semver2/system.drawing.common/index.json 290ms
  OK https://api.nuget.org/v3/registration5-gz-semver2/system.drawing.common/index.json 588ms
  OK https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/registrations2-semver2/system.drawing.common/index.json 782ms
  NotFound https://pkgs.dev.azure.com/azure-public/3ccf6661-f8ce-4e8a-bb2e-eff943ddd3c7/_packaging/be90bfa8-5c00-452b-ba9f-226ce4366992/nuget/v3/registrations2-semver2/system.drawing.common/index.json 983ms
  OK https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/7d9f5c21-0d79-403f-bfe3-9a4506529760/nuget/v3/registrations2-semver2/system.drawing.common/index.json 1603ms
Getting restore information for project /Users/user/Compile/mac-playground/System.Windows.Forms/System.Windows.Forms.csproj
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(109,5) : error NETSDK1139: The target platform identifier macos was not recognized.
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(109,5) : error NETSDK1139: The target platform identifier macos was not recognized.
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(109,5) : error NETSDK1139: The target platform identifier macos was not recognized.
Unable to create package spec for project. '/Users/user/Compile/mac-playground/System.Windows.Forms/System.Windows.Forms.csproj'
```

Microsoft.DotNet.ApiCompat
```
  GET https://pkgs.dev.azure.com/azure-public/3ccf6661-f8ce-4e8a-bb2e-eff943ddd3c7/_packaging/be90bfa8-5c00-452b-ba9f-226ce4366992/nuget/v3/registrations2-semver2/microsoft.dotnet.apicompat/index.json
  GET https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/7d9f5c21-0d79-403f-bfe3-9a4506529760/nuget/v3/registrations2-semver2/microsoft.dotnet.apicompat/index.json
  GET https://api.nuget.org/v3/registration5-gz-semver2/microsoft.dotnet.apicompat/index.json
  GET https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/registrations2-semver2/microsoft.dotnet.apicompat/index.json
  GET https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/registrations2-semver2/microsoft.dotnet.apicompat/index.json
  NotFound https://api.nuget.org/v3/registration5-gz-semver2/microsoft.dotnet.apicompat/index.json 587ms
  NotFound https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/registrations2-semver2/microsoft.dotnet.apicompat/index.json 718ms
  NotFound https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/7d9f5c21-0d79-403f-bfe3-9a4506529760/nuget/v3/registrations2-semver2/microsoft.dotnet.apicompat/index.json 851ms
  NotFound https://pkgs.dev.azure.com/azure-public/3ccf6661-f8ce-4e8a-bb2e-eff943ddd3c7/_packaging/be90bfa8-5c00-452b-ba9f-226ce4366992/nuget/v3/registrations2-semver2/microsoft.dotnet.apicompat/index.json 943ms
  OK https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/registrations2-semver2/microsoft.dotnet.apicompat/index.json 1242ms
Getting restore information for project /Users/user/Compile/mac-playground/System.Windows.Forms.Primitives/System.Windows.Forms.Primitives.csproj
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(109,5) : error NETSDK1139: The target platform identifier macos was not recognized.
Unable to create package spec for project. '/Users/user/Compile/mac-playground/System.Windows.Forms.Primitives/System.Windows.Forms.Primitives.csproj'
```
I'm new to this process on macos and can't figure it out how to proceed.
I'm using Visual Studio for Mac Community v8.9.2(build 0) on MacOS Big Sur 11.1 (for now).
Thank you!

**Comment by filipnavara** [comment]
The current state of the repository is that it needs .NET 6 Previews to build and even then it may need specific versions and workload packs.

You can edit this line in Directory.Build.props and set it to `true`:
https://github.com/emclient/mac-playground/blob/2a59d50203653c0640e4b3fd9d3642af926a95df/Directory.Build.props#L8
That should enable building just the regular classic Xamarin versions.

**Comment by badrianiulian** [comment]
I tried by setting the specified tag to `true`.
Now when opening the solution I get only one line regarding these errors but when running the FormsTest app, I still get errors regarding NuGet packages (in System.Private.CoreLib, System.Drawing.Common, MacApi and Mono.WebBrowser).

**Comment by filipnavara** [comment]
Sorry, took me a moment to look into it.

To completely disable the .NET 6 bits you need to change Directory.Build.props to:
```
<Project>
  <PropertyGroup>
    <MacDirectoryBuildPropsPath>$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)/.., '$(_DirectoryBuildPropsFile)'))</MacDirectoryBuildPropsPath>
  </PropertyGroup>
  <Import Project="$(MacDirectoryBuildPropsPath)\$(_DirectoryBuildPropsFile)" Condition="Exists('$(MacDirectoryBuildPropsPath)\$(_DirectoryBuildPropsFile)')"/>

  <PropertyGroup>
    <SkipMacProjects Condition="$(SkipMacProjects) == ''">false</SkipMacProjects>
    <SkipMacNet6Projects>true</SkipMacNet6Projects>
  </PropertyGroup>
</Project>
```

and global.json to:
```
{
  "msbuild-sdks": {
    "MSBuild.Sdk.Extras": "3.0.22"
  }
}
```

The .sln file is not up-to-date either since I was not building from Visual Studio for Mac for a long time. Adding existing project System.Drawing\System.Drawing.csproj should fix that. However there's still couple more error that require porting some workaround bits from our proprietary build files. :-/ I will try to find some time to do that.

**Comment by filipnavara** [comment]
This is enough to fix the build outside of Visual Studio:
https://github.com/emclient/mac-playground/commit/1e656443cfff4a34dd30b64283b649b38958e955

VS still fails to restore the packages properly or run the resulting executable from the correct location.

**Comment by badrianiulian** [comment]
So.. I modified the files as recommended.
The project loads fine and there are no errors regarding the NuGet packages.
But there is another error:
```
/usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(5,5): Error NETSDK1005: Assets file '/Users/user/Compile/mac-playground/Mono.WebBrowser/obj/project.assets.json' doesn't have a target for 'netstandard2.1'. Ensure that restore has run and that you have included 'netstandard2.1' in the TargetFrameworks for your project. (NETSDK1005) (Mono.WebBrowser)
```
I opened a Terminal and went to the root of the project (sln location):  `dotnet build` results in lots of red (errors).
Went in the Mono.WebBrowser folder: `dotnet build` -> `Mono.WebBrowser/bin/Debug/netstandard2.1/Mono.WebBrowser.dll Build succeeded.`

I tried modifying Mono.WebBrowser.csproj and changing TargetFramework to xamarinmac20 but it's a bust:
```
Line 30: /Users/user/Compile/mac-playground/Mono.WebBrowser/Mono.WebBrowser/IWebBrowser.cs(26,26): Error CS0234: The type or namespace name 'Specialized' does not exist in the namespace 'System.Collections' (are you missing an assembly reference?) (CS0234) (Mono.WebBrowser)
Line 205: /Users/user/Compile/mac-playground/Mono.WebBrowser/Mono.WebBrowser/IWebBrowser.cs(11,11): Error CS0246: The type or namespace name 'StringCollection' could not be found (are you missing a using directive or an assembly reference?) (CS0246) (Mono.WebBrowser)
Line 206: /Users/user/Compile/mac-playground/Mono.WebBrowser/Mono.WebBrowser/IWebBrowser.cs(11,11): Error CS0246: The type or namespace name 'StringCollection' could not be found (are you missing a using directive or an assembly reference?) (CS0246) (Mono.WebBrowser)
Line 254: /Users/user/Compile/mac-playground/Mono.WebBrowser/Mono.WebBrowser/IWebBrowser.cs(10,10): Error CS0246: The type or namespace name 'StringCollection' could not be found (are you missing a using directive or an assembly reference?) (CS0246) (Mono.WebBrowser)
Line 258: /Users/user/Compile/mac-playground/Mono.WebBrowser/Mono.WebBrowser/IWebBrowser.cs(10,10): Error CS0246: The type or namespace name 'StringCollection' could not be found (are you missing a using directive or an assembly reference?) (CS0246) (Mono.WebBrowser)
Line 347: /Users/user/Compile/mac-playground/Mono.WebBrowser/Mono.WebBrowser/IWebBrowser.cs(45,45): Error CS0234: The type or namespace name 'ComponentModel' does not exist in the namespace 'System' (are you missing an assembly reference?) (CS0234) (Mono.WebBrowser)
Line 419: /Users/user/Compile/mac-playground/Mono.WebBrowser/Mono.WebBrowser/IWebBrowser.cs(53,53): Error CS0234: The type or namespace name 'ComponentModel' does not exist in the namespace 'System' (are you missing an assembly reference?) (CS0234) (Mono.WebBrowser)
```

**Comment by filipnavara** [comment]
I am not quite sure what is going on with VS4Mac yet. Building Mono.WebBrowser from command line helped. After that VS4Mac was able to build most of the projects but it was not able to run FormsTest. I guess that may just be incompatibility with MSBuild.Sdk.Extras that's used for the projects.

The command line compilation has to be done with `msbuild` as opposed to `dotnet build` until .NET 6 comes along.

**Comment by badrianiulian** [comment]
So... after testing, the only setting that worked, involves modifying in Mono.WebBrowser.csproj this line:
```
-    <TargetFramework>netstandard2.1</TargetFramework>
+    <TargetFrameworks>netstandard2.0</TargetFrameworks>
```
Removed `bin` and `obj` after each modification while I was testing this. I opened the main solution each time so that the NuGet packages were restored. Closed Visual Studio for Mac and went into the Terminal once again but this time, I wrote `msbuild` as recomended and everything went just fine: compiled with 28 warnings.
Went into the mac-playground/FormsTest/bin/Debug/xamarinmac20/osx-x64/ folder and launched FormsTest.app and finally I saw the window of the FormsTest project.
It does have some bugs but it's working.
Thanks for the help and thank you for your work on this project!
Now I'll go on to testing it, using some other project.

**Comment by filipnavara** [comment]
I'm sorry for the crappy build experience. I hope to get it much cleaner with .NET 6 but so far that meant investing my time in Xamarin/.NET itself and neglecting this project a bit.

**Comment by badrianiulian** [comment]
When it comes to C#, .NET, Mono and Xamarin, I'm flying blind and only guided by my gut, so I understand your build experience problems. I enjoy playing more with the old C/C++ than having all this headaches for just building some visual cross platform programs.
Anyways, I found online that when compiling the project in VS2019(all 2019 versions i guess?), it helps getting rid of some #nullable errors by adding `<Nullable>annotations</Nullable>` after `<LangVersion>latest</LangVersion>` in all csproj files. So basically in System.Drawing.Common.csproj, System.Windows.Forms.Primitives.csproj and System.Windows.Forms.csproj. It's something related to using old style C# compared to C# 8.0 or something...

**Comment by filipnavara** [comment]
We use this code internally within a slightly different build system where we inherit some of the properties from Directory.Build.[props/targets] so we don't necessarily see all the same errors/warnings there.

Here's a little back story on the current situation. We are continually experiencing issues with the WinForms designer on .NET 5 (unrelated to this repository). One of the reasons is that lot of types were moved to different assemblies and the designer is incorrectly serialising the new names and version numbers of those assemblies into the .resx files. This is something that should be address with upcoming Visual Studio update which was not released yet. Since our project builds from the same code base on Windows and macOS we had to make type-forwarding shims to make the newly edited .resx files working. That worked for us for couple of months but it started to become unmaintainable. At that point we already had a .NET 5/6 branch of code in this repository where the layout of the projects was rearranged to more closely mirror what is used on Windows. Unfortunately .NET 6 is not going to be released until the end of the year. So what we did was to backport the .NET 6 branch of our code to target the legacy Xamarin through MSBuild.Sdk.Extras (which we already used elsewhere). This was all merged in a rush and synced into this repository right after .NET 6 Preview 1 was released. As much as I would to invest more time into cleaning up this repository I get pre-empted by more urgent things and it's left in slightly bit-rotten way. Note, however, that the issues are mainly with the build system while the code itself is used in production.
